eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -398,6 +405,20 @@ transValue (MaryValue n m) = justAda <> transMultiAsset m
 -- =============================================
 -- translate fileds like TxCert, Withdrawals, and similar
 
+unScriptPurposeV1 :: PlutusScriptPurpose 'PlutusV1 -> PV1.ScriptPurpose
+unScriptPurposeV1 (ScriptPurposePlutusV1 x) = x
+
+unScriptPurposeV2 :: PlutusScriptPurpose 'PlutusV2 -> PV2.ScriptPurpose
+unScriptPurposeV2 (ScriptPurposePlutusV2 x) = x
+
+unScriptPurposeV3 :: PlutusScriptPurpose 'PlutusV3 -> PV3.ScriptPurpose
+unScriptPurposeV3 (ScriptPurposePlutusV3 x) = x
+
+data PlutusScriptPurpose (l :: Language) where
+  ScriptPurposePlutusV1 :: PV1.ScriptPurpose -> PlutusScriptPurpose 'PlutusV1
+  ScriptPurposePlutusV2 :: PV2.ScriptPurpose -> PlutusScriptPurpose 'PlutusV2
+  ScriptPurposePlutusV3 :: PV3.ScriptPurpose -> PlutusScriptPurpose 'PlutusV3
```

Comment:
This is one of those times when it is safe to use record syntax with a sum type, just because it is a GADT:
```suggestion
data PlutusScriptPurpose (l :: Language) where
  ScriptPurposePlutusV1 :: 
    { unScriptPurposeV1 :: PV1.ScriptPurpose } -> PlutusScriptPurpose 'PlutusV1
  ScriptPurposePlutusV2 :: 
    { unScriptPurposeV2 :: PV2.ScriptPurpose } -> PlutusScriptPurpose 'PlutusV2
  ScriptPurposePlutusV3 :: 
    { unScriptPurposeV3 :: PV3.ScriptPurpose } -> PlutusScriptPurpose 'PlutusV3
```

Reply:
Great, thanks